File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
 public:
  bool searchMatrix(vector<vector<int>> &matrix, int target) {
    int row = matrix.size();
    if (row == 0) return false;
    int col = matrix[0].size();
    int left = 0;
    int right = row * col - 1;
    int pivot, pivot_element;
    while (left <= right) {
      pivot = (left + right) / 2;
      pivot_element = matrix[pivot / col][pivot % col];
      if (target == pivot_element) {
        return true;
      }
      if (target < pivot_element) {
        right = pivot - 1;
      } else {
        left = pivot + 1;
      }
    }
    return false;
  }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
# time complexity: O(logmn)
# space complexity: O(1)
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        ROW = len(matrix)
        COL = len(matrix[0])
        left = 0
        right = ROW * COL - 1
        if ROW == 0:
            return False
        while left <= right:
            mid = left + (right - left) // 2
            value = matrix[mid//COL][mid % COL]
            if value < target:
                left = mid + 1
            elif value > target:
                right = mid - 1
            else:
                return True
        return False
'''

0 1 2 3 4 5 6 7 8

0 1 2
3 4 5
6 7 8

row = idx // COL
col = idx % COL
'''

matrix = [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]]
target = 3
print(Solution().searchMatrix(matrix, target))
matrix = [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]]
target = 13
print(Solution().searchMatrix(matrix, target))
